lib/active_model/serializer/attributes.rb

```diff
@@ -1,55 +1,32 @@
 module ActiveModel
   class Serializer
     module Attributes
-      # @api private
-      class Attribute
-        delegate :call, to: :reader
-
-        attr_reader :name, :reader
-
-        def initialize(name)
-          @name = name
-          @reader = :no_reader
-        end
-
-        def self.build(name, block)
-          if block
-            AttributeBlock.new(name, block)
-          else
-            AttributeReader.new(name)
-          end
-        end
-      end
-      # @api private
-      class AttributeReader < Attribute
-        def initialize(name)
-          super(name)
-          @reader = ->(instance) { instance.read_attribute_for_serialization(name) }
-        end
-      end
-      # @api private
-      class AttributeBlock < Attribute
-        def initialize(name, block)
-          super(name)
-          @reader = ->(instance) { instance.instance_eval(&block) }
-        end
-      end
-
       extend ActiveSupport::Concern
 
       included do
         with_options instance_writer: false, instance_reader: false do |serializer|
-          serializer.class_attribute :_attribute_mappings # @api private : maps attribute key names to names to names of implementing methods, @see #attribute
-          self._attribute_mappings ||= {}
+          serializer.class_attribute :_attribute_procs # @api private : maps attribute key names to names to names of implementing methods, @see #attribute
+          self._attribute_procs ||= {}
+          serializer.class_attribute :_attribute_keys # @api private : maps attribute names to keys, @see #attribute
+          self._attribute_keys ||= {}
         end
 
         # Return the +attributes+ of +object+ as presented
         # by the serializer.
         def attributes(requested_attrs = nil, reload = false)
           @attributes = nil if reload
-          @attributes ||= self.class._attribute_mappings.each_with_object({}) do |(key, attribute_mapping), hash|
+          @attributes ||= self.class._attribute_keys.each_with_object({}) do |(name, key), hash|
             next unless requested_attrs.nil? || requested_attrs.include?(key)
-            hash[key] = attribute_mapping.call(self)
+            hash[key] = _attribute_value(name)
+          end
+        end
+
+        # @api private
+        def _attribute_value(name)
+          if self.class._attribute_procs[name]
+            instance_eval(&self.class._attribute_procs[name])
+          else
+            read_attribute_for_serialization(name)
```

Comment:
- so, I kind of like the polymorphism better.
- but at the end of the day, this is less code and the switch is entirely contained here, which is good
- on the other hand, now there's a data clump of ` _attribute_procs` and ` _attribute_keys` passed around
- and ` _attribute_keys` sounds a lot like `_attributes_keys`, which seems to be a bug attractor to me
- but `_attribute_value(name)` is a nicer interface than `attribute_mapping.call(self)` and is more consistent with the reflection value

So, how about

```ruby
class Attribute
  attr_reader :name, :block
  def initialize(attr, block = nil)
    @name = attr
    @block = block
  end
  def value(instance)
    if @block
      instance.instance_eval(&block)
    else
      instance.read_attribute_for_serialization(name) 
    end
end
```

and we've removed storing the unnecessary lambdas, introduced a smaller change to the serializer interface, and isolated the attribute vs. block switch with a consistent interface to Reflection?

```ruby
         def attribute(attr, options = {}, &block)
           key = options.fetch(:key, attr)
           _attribute_mappings[key] = Attribute.new(attr, block)
```